File: src/Fable.ReactAgGrid.fs
module Fable.ReactAgGrid

open Browser.Types
open Fable.Core
open Fable.Core.JsInterop
open Fable.React

[<RequireQualifiedAccess>]
module Grid =

    type ColDef =
        { headerName : string
          field : string }

    type RowNode =
        { id : string }

    type Column =
        { id : string }

    type Event =
        { data : obj
          newValue : obj
          oldValue : obj
          colDef: ColDef
          column: Column
          node : RowNode }

    type CellEditorParams =
        { value : obj
          keyPress : float
          charPress : string
          column: Column;
          node: RowNode;
          rowIndex: float

        //   api: GridApi; TODO

          ///   columnApi: ColumnApi; TODO
          cellStartedEdit: bool;
          // the grid's context object
          context: obj;

          /// callback to tell grid a key was pressed - useful to pass control key events (tab, arrows etc)
          /// back to grid - however you do
          onKeyDown: (KeyboardEvent) -> unit;

          /// Callback to tell grid to stop editing the current cell. pass 'false' to prevent navigation moving
          /// to the next cell if grid property enterMovesDownAfterEdit=true
          stopEditing: (bool) -> unit;

          /// A reference to the DOM element representing the grid cell that your component will live inside. Useful if you
          /// want to add event listeners or classes at this level. This is the DOM element that gets browser focus when selecting cells.
          eGridCell: HTMLElement;

          /// Utility function to parse a value using the column's colDef.valueParser
          parseValue: (obj) -> obj;

          // Utility function to format a value using the column's colDef.valueFormatter
          formatValue: (obj) -> obj;}

    type CellClassParams = {
        // The value to be rendered.
        value: obj
        // The row (from the rowData array, where value was taken) been rendered.
        data: obj
        // The node associated to this row
        node: RowNode
        // The colDef been rendered
        colDef: ColDef
        // The index of the row about to be rendered
        rowIndex: int
        // A reference to the ag-Grid API.
        // api: GridApi
        // If provided in gridOptions, a context object
        context: obj
    }

    type RowSpanParams = {
        node: obj // row node in question
        data: RowNode // data for the row
        colDef: ColDef // the col def for the column
        column: Column // the column object in question
        //api: GridApi // the grid's API // TODO
        //columnApi: ColumnApi // the grid's column API // TODO
        context: obj // the provided context
    }

    type ColumnDefOptions =
        | HeaderName of string
        | Children of obj array
        | ColumnGroupShow of bool
        | HeaderClass of obj
        | ToolPanelClass of obj
        | SuppressColumnsToolPanel of bool
        | SuppressFiltersToolPanel of bool
        | Field of string
        | ColId of string
        | Type of string array
        | Width of int
        | Height of int
        | MinWidth of int
        | MaxWidth of int

        ///TODO: Descrip
        | Flex of int
        | Filter of bool
        | FilterParams of obj // TODO
        | FloatingFilterComponent of obj // TODO
        | FloatingFilterComponentParams of obj // TODO
        | Hide of bool
        | Pinned of string
        | LockPosition of bool
        | LockVisible of bool
        | LockPinned of bool
        | Sortable of bool
        | Sort of string option
        | SortedAt of int // TODO
        | SortingOrder of string option
        | Resizable of bool
        | HeaderTooltip of obj //TODO
        | TooltipField of obj //TODO
        | CheckboxSelection of bool
        | RowDrag of bool //TODO enable functions to be added
        | DndSource of bool //TODO enable functions to be added
        | Editable of bool
        | CellStyle of (CellClassParams -> obj)
        | CellClass of (CellClassParams -> obj)
        | AutoHeight of bool
        | SingleClickEdit of bool
        | OnCellValueChanged of (Event -> unit)
        | RowSpan of (RowSpanParams -> obj)
        // TODO add remaining options

    let ColumnDef options =
        keyValueList CaseRules.LowerFirst options

    type GridOptions =
        | ColumnDefs of obj array
        | DefaultColDef of obj
        | PinnedTopRowData  of obj array
        | PinnedBottomRowData  of obj array
        | RowData of obj array
        | RowHeight of float
        | HeaderHeight of float
        | RowStyle of obj
        | StopEditingWhenGridLosesFocus of bool
        | SuppressRowTransform of bool

    let inline grid (props : GridOptions list) (elems : ReactElement list) : ReactElement =
        ofImport "AgGridReact" "ag-grid-react" (keyValueList CaseRules.LowerFirst props) elems

File: app/output/src/Fable.ReactAgGrid.js
import { Union, Record } from "../.fable/fable-library.3.0.0-nagareyama-rc-005/Types.js";
import { union_type, option_type, array_type, int32_type, lambda_type, unit_type, class_type, bool_type, float64_type, obj_type, record_type, string_type } from "../.fable/fable-library.3.0.0-nagareyama-rc-005/Reflection.js";
import { keyValueList } from "../.fable/fable-library.3.0.0-nagareyama-rc-005/MapUtil.js";

export class ColDef extends Record {
    constructor(headerName, field) {
        super();
        this.headerName = headerName;
        this.field = field;
    }
}

export function ColDef$reflection() {
    return record_type("Fable.ReactAgGrid.Grid.ColDef", [], ColDef, () => [["headerName", string_type], ["field", string_type]]);
}

export class RowNode extends Record {
    constructor(id, data) {
        super();
        this.id = id;
        this.data = data;
    }
}

export function RowNode$reflection() {
    return record_type("Fable.ReactAgGrid.Grid.RowNode", [], RowNode, () => [["id", string_type], ["data", string_type]]);
}

export class Column extends Record {
    constructor(id) {
        super();
        this.id = id;
    }
}

export function Column$reflection() {
    return record_type("Fable.ReactAgGrid.Grid.Column", [], Column, () => [["id", string_type]]);
}

export class Event$ extends Record {
    constructor(data, newValue, oldValue, colDef, column, node) {
        super();
        this.data = data;
        this.newValue = newValue;
        this.oldValue = oldValue;
        this.colDef = colDef;
        this.column = column;
        this.node = node;
    }
}

export function Event$$reflection() {
    return record_type("Fable.ReactAgGrid.Grid.Event", [], Event$, () => [["data", obj_type], ["newValue", obj_type], ["oldValue", obj_type], ["colDef", ColDef$reflection()], ["column", Column$reflection()], ["node", RowNode$reflection()]]);
}

export class CellEditorParams extends Record {
    constructor(value, keyPress, charPress, column, node, rowIndex, cellStartedEdit, context, onKeyDown, stopEditing, eGridCell, parseValue, formatValue) {
        super();
        this.value = value;
        this.keyPress = keyPress;
        this.charPress = charPress;
        this.column = column;
        this.node = node;
        this.rowIndex = rowIndex;
        this.cellStartedEdit = cellStartedEdit;
        this.context = context;
        this.onKeyDown = onKeyDown;
        this.stopEditing = stopEditing;
        this.eGridCell = eGridCell;
        this.parseValue = parseValue;
        this.formatValue = formatValue;
    }
}

export function CellEditorParams$reflection() {
    return record_type("Fable.ReactAgGrid.Grid.CellEditorParams", [], CellEditorParams, () => [["value", obj_type], ["keyPress", float64_type], ["charPress", string_type], ["column", Column$reflection()], ["node", RowNode$reflection()], ["rowIndex", float64_type], ["cellStartedEdit", bool_type], ["context", obj_type], ["onKeyDown", lambda_type(class_type("Browser.Types.KeyboardEvent"), unit_type)], ["stopEditing", lambda_type(bool_type, unit_type)], ["eGridCell", class_type("Browser.Types.HTMLElement")], ["parseValue", lambda_type(obj_type, obj_type)], ["formatValue", lambda_type(obj_type, obj_type)]]);
}

export class CellClassParams extends Record {
    constructor(value, data, node, colDef, rowIndex, context) {
        super();
        this.value = value;
        this.data = data;
        this.node = node;
        this.colDef = colDef;
        this.rowIndex = (rowIndex | 0);
        this.context = context;
    }
}

export function CellClassParams$reflection() {
    return record_type("Fable.ReactAgGrid.Grid.CellClassParams", [], CellClassParams, () => [["value", obj_type], ["data", obj_type], ["node", RowNode$reflection()], ["colDef", ColDef$reflection()], ["rowIndex", int32_type], ["context", obj_type]]);
}

export class RowSpanParams extends Record {
    constructor(node, data, colDef, column, context) {
        super();
        this.node = node;
        this.data = data;
        this.colDef = colDef;
        this.column = column;
        this.context = context;
    }
}

export function RowSpanParams$reflection() {
    return record_type("Fable.ReactAgGrid.Grid.RowSpanParams", [], RowSpanParams, () => [["node", obj_type], ["data", RowNode$reflection()], ["colDef", ColDef$reflection()], ["column", Column$reflection()], ["context", obj_type]]);
}

export class ColumnDefOptions extends Union {
    constructor(tag, ...fields) {
        super();
        this.tag = (tag | 0);
        this.fields = fields;
    }
    cases() {
        return ["HeaderName", "Children", "ColumnGroupShow", "HeaderClass", "ToolPanelClass", "SuppressColumnsToolPanel", "SuppressFiltersToolPanel", "Field", "ColId", "Type", "Width", "Height", "MinWidth", "MaxWidth", "Flex", "Filter", "FilterParams", "FloatingFilterComponent", "FloatingFilterComponentParams", "Hide", "Pinned", "LockPosition", "LockVisible", "LockPinned", "Sortable", "Sort", "SortedAt", "SortingOrder", "Resizable", "HeaderTooltip", "TooltipField", "CheckboxSelection", "RowDrag", "DndSource", "Editable", "CellStyle", "CellClass", "AutoHeight", "SingleClickEdit", "OnCellValueChanged", "RowSpan"];
    }
}

export function ColumnDefOptions$reflection() {
    return union_type("Fable.ReactAgGrid.Grid.ColumnDefOptions", [], ColumnDefOptions, () => [[["Item", string_type]], [["Item", array_type(obj_type)]], [["Item", bool_type]], [["Item", obj_type]], [["Item", obj_type]], [["Item", bool_type]], [["Item", bool_type]], [["Item", string_type]], [["Item", string_type]], [["Item", array_type(string_type)]], [["Item", int32_type]], [["Item", int32_type]], [["Item", int32_type]], [["Item", int32_type]], [["Item", int32_type]], [["Item", bool_type]], [["Item", obj_type]], [["Item", obj_type]], [["Item", obj_type]], [["Item", bool_type]], [["Item", string_type]], [["Item", bool_type]], [["Item", bool_type]], [["Item", bool_type]], [["Item", bool_type]], [["Item", option_type(string_type)]], [["Item", int32_type]], [["Item", option_type(string_type)]], [["Item", bool_type]], [["Item", obj_type]], [["Item", obj_type]], [["Item", bool_type]], [["Item", bool_type]], [["Item", bool_type]], [["Item", bool_type]], [["Item", lambda_type(CellClassParams$reflection(), obj_type)]], [["Item", lambda_type(CellClassParams$reflection(), obj_type)]], [["Item", bool_type]], [["Item", bool_type]], [["Item", lambda_type(Event$$reflection(), unit_type)]], [["Item", lambda_type(RowSpanParams$reflection(), int32_type)]]]);
}

export function ColumnDef(options) {
    return keyValueList(options, 1);
}

export class GridOptions extends Union {
    constructor(tag, ...fields) {
        super();
        this.tag = (tag | 0);
        this.fields = fields;
    }
    cases() {
        return ["ColumnDefs", "DefaultColDef", "RowData", "PinnedTopRowData", "PinnedBottomRowData", "RowHeight", "HeaderHeight", "RowStyle", "StopEditingWhenGridLosesFocus", "SuppressRowTransform"];
    }
}

export function GridOptions$reflection() {
    return union_type("Fable.ReactAgGrid.Grid.GridOptions", [], GridOptions, () => [[["Item", array_type(obj_type)]], [["Item", obj_type]], [["Item", array_type(obj_type)]], [["Item", array_type(obj_type)]], [["Item", array_type(obj_type)]], [["Item", float64_type]], [["Item", float64_type]], [["Item", obj_type]], [["Item", bool_type]], [["Item", bool_type]]]);
}

